pkg/verifier/verifier.go

```diff
@@ -0,0 +1,147 @@
+	"context"
+	"sync"
+
+	"github.com/rotisserie/eris"
+	"github.com/solo-io/go-utils/contextutils"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+)
+
+// ServerVerifyOption is used to specify one of several options on
+// whether and how to verify the resource's availabiltiy on the API server
+type ServerVerifyOption int
+
+const (
+	// skip verifying whether a resource is  for a kind before reading it from the API server
+	ServerVerifyOption_SkipVerify ServerVerifyOption = iota
+
+	// return an error if the resource does not exist for a kind before reading it from the API server
+	ServerVerifyOption_ErrorIfNotPresent
+
+	// log an error (and continue) if the resource does not exist for a kind before reading it from the API server
+	// the reconcile loop will not be started if the server resource is not supported.
+	ServerVerifyOption_WarnIfNotPresent
+
+	// ignore error (and continue) if the resource does not exist for a kind before reading it from the API server.
+	// the reconcile loop will not be started if the server resource is not supported.
+	ServerVerifyOption_IgnoreIfNotPresent
+)
+
+// ServerResourceVerifier verifies whether a given cluster server supports a given resource.
+type ServerResourceVerifier interface {
+	// Verify whether the API server for the given rest.Config supports the resource with the given GVK.
+	// For the "local/management" cluster, set cluster to ""
+	// Note that once a resource has been verified, the result will be cached for subsequent calls.
+	// Returns true if the resource is registered, false otherwise.
+	// an error will be returned if the ErrorIfNotPresent option is selected for the given GVK
+	VerifyServerResource(cluster string, cfg *rest.Config, gvk schema.GroupVersionKind) (bool, error)
+}
+
+type verifier struct {
+	// used for logging
+	ctx context.Context
+
+	// set of per-resource-type verify options; default is Skip.
+	options map[schema.GroupVersionKind]ServerVerifyOption
+
+	// set when a resource is successfully verified for the first time.
+	// future calls to VerifyServerResource will return quickly if the
+	// resources have already been verified for the cluster.
+	verifiedClusterResources map[string]map[schema.GroupVersionKind]bool
+	lock                     sync.RWMutex
+}
+
+func NewVerifier(
+	ctx context.Context,
+	options map[schema.GroupVersionKind]ServerVerifyOption,
+) *verifier {
+	if options == nil {
+		options = map[schema.GroupVersionKind]ServerVerifyOption{}
+	}
+	return &verifier{
+		ctx:                      ctx,
+		options:                  options,
+		verifiedClusterResources: map[string]map[schema.GroupVersionKind]bool{},
+	}
+}
+
+// Verify whether the API server for the given rest.Config supports the resource with the given GVK.
+func (v *verifier) VerifyServerResource(cluster string, cfg *rest.Config, gvk schema.GroupVersionKind) (bool, error) {
+	v.lock.RLock()
```

Comment:
The use of locks here is a bit hard to reason about, could we possibly move the map reads and writes to their own functions for clarity?